src/cargo/ops/registry/mod.rs

```diff
@@ -186,12 +191,35 @@ fn registry(
 /// crates.io (such as index.crates.io), while the second is always the original source.
 fn get_source_id(
     gctx: &GlobalContext,
+    package_id: Option<PackageId>,
     reg_or_index: Option<&RegistryOrIndex>,
-) -> CargoResult<RegistrySourceIds> {
-    let sid = match reg_or_index {
-        None => SourceId::crates_io(gctx)?,
-        Some(RegistryOrIndex::Index(url)) => SourceId::for_registry(url)?,
-        Some(RegistryOrIndex::Registry(r)) => SourceId::alt_registry(gctx, r)?,
+) -> CargoResult<(bool, RegistrySourceIds)> {
+    let (use_package_source_id, sid) = match (&reg_or_index, package_id) {
```

Comment:
Also not too thrilled with this.  I haven't dug into this but is there a way we can split this into layers?  If not, should we just keep these as two functions?